confluent_kafka/avro/serializer/__init__.py

```diff
@@ -33,8 +61,270 @@ def __str__(self):
+        :returns: decoder function
+        :rtype: func
+        """
+
+        if schema_id not in self.codec_cache:
+            try:
+                schema = self.registry_client.get_by_id(schema_id)
+                if not schema:
+                    raise SerializerError("Schema does not exist", self.is_key)
+                self.codec_cache[schema_id] = self._get_encoder_func(schema)
+            except ClientError:
+                exc_type, exc_value, exc_traceback = sys.exc_info()
+                raise SerializerError(
+                    repr(traceback.format_exception(exc_type, exc_value, exc_traceback)),
+                    self.is_key)
+
+        # get the writer
+        writer = self.codec_cache[schema_id]
+        with ContextStringIO() as outf:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            outf.write(struct.pack('>bI', MAGIC_BYTE, schema_id))
+
+            # write the record to the rest of the buffer
+            writer(datum, outf)
+            return outf.getvalue()
+
+
+class AvroDeserializer(object):
+    """
+    Decodes Kafka messages encoded by Confluent Schema Registry compliant Avro Serializers.
+
+    :param registry_client CachedSchemaRegistryClient: Instance of CachedSchemaRegistryClient.
+    :param bool is_key: True if configured as a key_serializer.
+    :param schema reader_schema: Optional reader schema to be used during deserialization.
+    """
+    __slots__ = ["registry_client", "codec_cache", "is_key", "reader_schema"]
+
+    def __init__(self, registry_client, is_key=False, reader_schema=None):
+        self.registry_client = registry_client
+        self.codec_cache = {}
+        self.is_key = is_key
+        self.reader_schema = reader_schema
+
+    def __call__(self, topic, datum):
+        """
+        Decode a datum from kafka that has been encoded for use with the Confluent Schema Registry.
+        :param str|bytes or None datum: message key or value to be decoded.
+        :returns: Decoded message key or value contents.
+        :rtype GenericAvroRecord:
+        """
+
+        if datum is None:
+            return None
+
+        if len(datum) <= 5:
+            raise SerializerError("message is too small to decode")
+
+        with ContextStringIO(datum) as payload:
+            magic, schema_id = struct.unpack('>bI', payload.read(5))
+            if magic != MAGIC_BYTE:
+                raise SerializerError("message does not start with magic byte", self.is_key)
+
+            decoder_func = self._get_decoder_func(schema_id, payload)
+            return decoder_func(payload)
+
+    def _get_decoder_func(self, schema_id, payload):
+        if schema_id in self.codec_cache:
+            return self.codec_cache[schema_id]
+
+        try:
+            writer_schema = self.registry_client.get_by_id(schema_id)
+        except ClientError as e:
+            raise SerializerError("unable to fetch schema with id %d: %s" % (schema_id, str(e)))
```

Comment:
I think this is throwing away useful information - users may want to handle connectivity issues differently than data serialization errors, so it would be good to be able to catch on the different types. I think it's ok to just let the `ClientError` bubble up. i.e. probably going over-board to invent a special type of serializer error for connectivity issues.

Reply:
ack